plugins/findhistory.koplugin/main.lua

```diff
@@ -0,0 +1,114 @@
+local ConfirmBox = require("ui/widget/confirmbox")
+local DataStorage = require("datastorage")
+local InfoMessage = require("ui/widget/infomessage")
+local MultiConfirmBox = require("ui/widget/multiconfirmbox")
+local ReadHistory = require("readhistory")
+local T = require("ffi/util").template
+local UIManager = require("ui/uimanager")
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
+local joinPath = require("ffi/util").joinPath
+local lfs = require("libs/libkoreader-lfs")
+local logger = require("logger")
+local _ = require("gettext")
+
+local menuItem = {
+    text = _("Retrieve reading records")
+}
+
+local history_file = joinPath(DataStorage:getDataDir(), "history.lua")
+local history_backup_file = joinPath(DataStorage:getDataDir(), "history.lua.backup")
+
+local function getFilePathFromMetadata(file)
+    local noSuffix, suffix = file:match("(.*)%.sdr/metadata%.(.+)%.lua")
+    return noSuffix .. "." .. suffix
+end
+
+local function doBuildHistory()
+    --- home_dir should be returned by some shared logic.
+    local dir = G_reader_settings:readSetting("home_dir")
+                or require("apps/filemanager/filemanagerutil").getDefaultDir()
+                or "."
+    local file = io.popen("find '" .. dir .. "' -name 'metadata.*.lua' -exec stat -c '%N %Y' {} \\;")
```

Comment:
Ouch.

LFS directory walking instead, maybe?

(c.f. PathChooser).

Reply:
Any reason to prefer lfs? It uses stat under the hood indeed.

Reply:
At the very least lfs is more portable. E.g., will it work on super ancient BusyBox/Android, it will definitely not work on Windows, etc.

Reply:
Ooops, I thought windows is not expected to be supported. Will update this part anyway.

Reply:
It's not "expected" but there's no reason to make it harder. ;-)